wallet-next/src/storage.rs

```diff
@@ -132,6 +134,86 @@ impl Storage {
         Ok(bincode::deserialize(result.bytes.as_slice())?)
     }
 
+    pub async fn notes(
+        &self,
+        include_spent: bool,
+        asset_id: Option<asset::Id>,
+        diversifier_index: Option<penumbra_crypto::keys::DiversifierIndex>,
+        amount_to_spend: u64,
+    ) -> anyhow::Result<Vec<NoteRecord>> {
+        // If set, return spent notes as well as unspent notes.
+        // bool include_spent = 2;
+        let spent_clause = match include_spent {
+            false => "NULL",
+            true => "height_spent",
+        };
+
+        // If set, only return notes with the specified asset id.
+        // crypto.AssetId asset_id = 3;
+
+        let asset_clause = match asset_id {
+            Some(x) => format!("{:?}", x),
+            None => "asset_id".to_string(),
+        };
+
+        // If set, only return notes with the specified diversifier index.
+        // crypto.DiversifierIndex diversifier_index = 4;
+
+        let diversifier_clause = match diversifier_index {
+            Some(x) => format!("{:?}", x),
+            None => "diversifier_index".to_string(),
+        };
+
+        // If set, stop returning notes once the total exceeds this amount.
+        //
+        // Ignored if `asset_id` is unset or if `include_spent` is set.
+        // uint64 amount_to_spend = 5;
+
+        if asset_id.is_none() || include_spent {
+            //TODO: figure out a clever way to only return notes up to the sum using SQL
```

Comment:
I don't know a succinct/straightforward way in SQL to truncate the size of the returned set by the sum of a column value, but if we refactored to stream smaller sets of `NoteRecord`s we could just sum those smaller sets and check against the `amount_to_spend` - as it stands, it's just returning the whole set of `NoteRecord`s at once, so to implement this cutoff, we'd need that to change.

Reply:
For now I think it's fine to load everything in memory and then throw it away. We can always do the better DB query if/when we notice a perf problem.